manifold/src/csg_tree.cpp

```diff
@@ -0,0 +1,447 @@
+    std::push_heap(results.begin(), results.end(), cmpFn);
+  }
+}
+
+/**
+ * Efficient union operation on a set of nodes by doing Compose as much as
+ * possible.
+ */
+void CsgOpNode::BatchUnion() const {
+  // INVARIANT: children_ is a vector of leaf nodes
+  // this kMaxUnionSize is a heuristic to avoid the pairwise disjoint check
+  // with O(n^2) complexity to take too long.
+  // If the number of children exceeded this limit, we will operate on chunks
+  // with size kMaxUnionSize.
+  constexpr int kMaxUnionSize = 1000;
+  while (children_.size() > 1) {
+    int start;
+    if (children_.size() > kMaxUnionSize) {
+      start = children_.size() - kMaxUnionSize;
+    } else {
+      start = 0;
+    }
+    VecDH<Box> boxes;
+    boxes.reserve(children_.size() - start);
+    for (int i = start; i < children_.size(); i++) {
+      boxes.push_back(std::dynamic_pointer_cast<CsgLeafNode>(children_[i])
+                          ->GetBoundingBox());
+    }
+    const Box *boxesD = boxes.cptrD();
+    // partition the children into a set of disjoint sets
+    // each set contains a set of children that are pairwise disjoint
+    std::vector<VecDH<size_t>> disjointSets;
+    for (size_t i = 0; i < boxes.size(); i++) {
+      auto lambda = [boxesD, i](const VecDH<size_t> &set) {
+        return find_if<decltype(set.end())>(
+                   autoPolicy(set.size()), set.begin(), set.end(),
+                   CheckOverlap({boxesD, i})) == set.end();
+      };
+      decltype(disjointSets.end()) it =
+          std::find_if(disjointSets.begin(), disjointSets.end(), lambda);
+      if (it == disjointSets.end()) {
+        disjointSets.push_back(std::vector<size_t>{i});
+      } else {
+        it->push_back(i);
+      }
+    }
+    // compose each set of disjoint children
+    std::vector<std::shared_ptr<const Manifold::Impl>> impls;
+    for (const auto &set : disjointSets) {
+      if (set.size() == 1) {
+        impls.push_back(
+            std::dynamic_pointer_cast<CsgLeafNode>(children_[start + set[0]])
+                ->GetImpl());
+      } else {
+        std::vector<std::shared_ptr<CsgLeafNode>> tmp;
+        for (size_t j : set) {
+          tmp.push_back(
+              std::dynamic_pointer_cast<CsgLeafNode>(children_[start + j]));
+        }
+        impls.push_back(
+            std::make_shared<const Manifold::Impl>(CsgLeafNode::Compose(tmp)));
+      }
+    }
+    BatchBoolean(Manifold::OpType::ADD, impls);
+    children_.erase(children_.begin() + start, children_.end());
+    children_.push_back(std::make_shared<CsgLeafNode>(impls.back()));
+    // move it to the front as we process from the back, and the newly added
+    // child should be quite complicated
+    std::swap(children_.front(), children_.back());
+  }
+}
+
+/**
+ * Flatten the children to a list of leaf nodes and return them.
+ * If finalize is true, the list will be guaranteed to be a list of leaf nodes
+ * (i.e. no ops). Otherwise, the list may contain ops.
```

Comment:
This seems odd; it says it will flatten the list, but if finalize is not true, then it might not? I try to avoid functions taking boolean inputs if I can, but maybe this is just an issue with the comment?

Reply:
The idea is that if `finalize` is false, we will try to flatten the tree if it is not expensive (if we don't have to evaluate the boolean expressions), and we will perform the evaluation if `finalize` is true. But yes, this is a bit confusing.

Actually this comment is outdated because we don't do any reordering for now, so `finalize` is kind of useless.

Reply:
Originally it will call this function with `finalize=false` when constructing the CSG tree, so unions will be flattened into a large union node with many children, which improves performance a bit. This is not necessary though.

Reply:
Okay, well let's deal with this in the next PR.